File: test/test_xwindowattributes.c
#include <X11/Xlib.h>
#include <caml/memory.h>
#include <caml/mlvalues.h>

extern value Val_xwindowattributes(XWindowAttributes *attributes);

// 1. Basic: A typical mapped InputOutput window with standard attributes
CAMLprim value get_xwa_basic(value unit) {
  CAMLparam1(unit);
  XWindowAttributes attr = {.x = 100,
                            .y = 200,
                            .width = 800,
                            .height = 600,
                            .border_width = 2,
                            .depth = 24,
                            .visual = (Visual *)0x12345678,
                            .root = 1234,
                            .class = InputOutput,
                            .bit_gravity = NorthWestGravity,
                            .win_gravity = NorthWestGravity,
                            .backing_store = NotUseful,
                            .backing_planes = 0xFFFFFFFF,
                            .backing_pixel = 0,
                            .save_under = False,
                            .colormap = 5678,
                            .map_installed = True,
                            .map_state = IsViewable,
                            .all_event_masks =
                                StructureNotifyMask | ExposureMask,
                            .your_event_mask = StructureNotifyMask,
                            .do_not_propagate_mask = 0,
                            .override_redirect = False,
                            .screen = (Screen *)0x87654321};
  CAMLreturn(Val_xwindowattributes(&attr));
}

// 2. Zeroed: All fields zeroed out, pointers set to NULL (Testing null
// references)
CAMLprim value get_xwa_zeroed(value unit) {
  CAMLparam1(unit);
  XWindowAttributes attr = {.x = 0,
                            .y = 0,
                            .width = 0,
                            .height = 0,
                            .border_width = 0,
                            .depth = 0,
                            .visual = (Visual *)0,
                            .root = 0,
                            .class = 0,
                            .bit_gravity = ForgetGravity, // 0
                            .win_gravity = UnmapGravity,  // 0
                            .backing_store = NotUseful,   // 0
                            .backing_planes = 0,
                            .backing_pixel = 0,
                            .save_under = False,
                            .colormap = 0,
                            .map_installed = False,
                            .map_state = IsUnmapped, // 0
                            .all_event_masks = 0,
                            .your_event_mask = 0,
                            .do_not_propagate_mask = 0,
                            .override_redirect = False,
                            .screen = (Screen *)0};
  CAMLreturn(Val_xwindowattributes(&attr));
}

// 3. Negative Coordinates: Placed out-of-bounds (Testing signed ints for
// coordinates)
CAMLprim value get_xwa_negative_coords(value unit) {
  CAMLparam1(unit);
  XWindowAttributes attr = {.x = -1500,
                            .y = -2000,
                            .width = 1920,
                            .height = 1080,
                            .border_width = 0,
                            .depth = 24,
                            .visual = (Visual *)0x1234,
                            .root = 1234,
                            .class = InputOutput,
                            .bit_gravity = NorthWestGravity,
                            .win_gravity = NorthWestGravity,
                            .backing_store = NotUseful,
                            .backing_planes = ~0UL,
                            .backing_pixel = 0,
                            .save_under = False,
                            .colormap = 5678,
                            .map_installed = True,
                            .map_state = IsViewable,
                            .all_event_masks = 0,
                            .your_event_mask = 0,
                            .do_not_propagate_mask = 0,
                            .override_redirect = False,
                            .screen = (Screen *)0x1234};
  CAMLreturn(Val_xwindowattributes(&attr));
}

// 4. Large Dimensions: Extremely large dimensions (Testing signed int
// boundaries)
CAMLprim value get_xwa_huge_dimensions(value unit) {
  CAMLparam1(unit);
  XWindowAttributes attr = {.x = 0,
                            .y = 0,
                            .width = 2147483647,
                            .height = 2147483647, // INT_MAX
                            .border_width = 2147483647,
                            .depth = 32,
                            .visual = (Visual *)0x1234,
                            .root = 123,
                            .class = InputOutput,
                            .bit_gravity = NorthWestGravity,
                            .win_gravity = NorthWestGravity,
                            .backing_store = NotUseful,
                            .backing_planes = ~0UL,
                            .backing_pixel = 0,
                            .save_under = False,
                            .colormap = 567,
                            .map_installed = True,
                            .map_state = IsViewable,
                            .all_event_masks = 0,
                            .your_event_mask = 0,
                            .do_not_propagate_mask = 0,
                            .override_redirect = False,
                            .screen = (Screen *)0x1234};
  CAMLreturn(Val_xwindowattributes(&attr));
}

// 5. Input Only: Invisible windows used for catching events (Depth 0, No
// Visual)
CAMLprim value get_xwa_input_only(value unit) {
  CAMLparam1(unit);
  XWindowAttributes attr = {.x = 10,
                            .y = 10,
                            .width = 100,
                            .height = 100,
                            .border_width = 0,
                            .depth = 0, // InputOnly windows always have depth 0
                            .visual = NULL, // No visual associated
                            .root = 999,
                            .class = InputOnly,
                            .bit_gravity = ForgetGravity,
                            .win_gravity = NorthWestGravity,
                            .backing_store = NotUseful,
                            .backing_planes = 0,
                            .backing_pixel = 0,
                            .save_under = False,
                            .colormap = None,
                            .map_installed = False,
                            .map_state = IsUnmapped,
                            .all_event_masks =
                                EnterWindowMask | LeaveWindowMask,
                            .your_event_mask = EnterWindowMask,
                            .do_not_propagate_mask = 0,
                            .override_redirect = True,
                            .screen = (Screen *)0x8888};
  CAMLreturn(Val_xwindowattributes(&attr));
}

// 6. Max Masks: Unsigned long limits for planes and masks
CAMLprim value get_xwa_max_masks_planes(value unit) {
  CAMLparam1(unit);
  XWindowAttributes attr = {.x = 0,
                            .y = 0,
                            .width = 10,
                            .height = 10,
                            .border_width = 0,
                            .depth = 24,
                            .visual = (Visual *)0x1111,
                            .root = 111,
                            .class = InputOutput,
                            .bit_gravity = NorthWestGravity,
                            .win_gravity = NorthWestGravity,
                            .backing_store = NotUseful,
                            .backing_planes = ~0UL, // Max unsigned long
                            .backing_pixel = ~0UL,  // Max unsigned long
                            .save_under = False,
                            .colormap = 111,
                            .map_installed = True,
                            .map_state = IsViewable,
                            .all_event_masks = ~0L, // Max long
                            .your_event_mask = ~0L,
                            .do_not_propagate_mask = ~0L,
                            .override_redirect = False,
                            .screen = (Screen *)0x1111};
  CAMLreturn(Val_xwindowattributes(&attr));
}

// 7. Bools True: Testing OCaml Bool mapping for all boolean fields
CAMLprim value get_xwa_bools_true(value unit) {
  CAMLparam1(unit);
  XWindowAttributes attr = {.x = 5,
                            .y = 5,
                            .width = 50,
                            .height = 50,
                            .border_width = 1,
                            .depth = 8,
                            .visual = (Visual *)0x2222,
                            .root = 222,
                            .class = InputOutput,
                            .bit_gravity = NorthWestGravity,
                            .win_gravity = NorthWestGravity,
                            .backing_store = NotUseful,
                            .backing_planes = 0xFF,
                            .backing_pixel = 0xFF,
                            .save_under = True, // Bool #1
                            .colormap = 222,
                            .map_installed = True, // Bool #2
                            .map_state = IsViewable,
                            .all_event_masks = 0,
                            .your_event_mask = 0,
                            .do_not_propagate_mask = 0,
                            .override_redirect = True, // Bool #3
                            .screen = (Screen *)0x2222};
  CAMLreturn(Val_xwindowattributes(&attr));
}

// 8. Odd Gravities & Backing Store (Testing enum variants)
CAMLprim value get_xwa_gravities_store(value unit) {
  CAMLparam1(unit);
  XWindowAttributes attr = {.x = 20,
                            .y = 20,
                            .width = 200,
                            .height = 200,
                            .border_width = 5,
                            .depth = 32,
                            .visual = (Visual *)0x3333,
                            .root = 333,
                            .class = InputOutput,
                            .bit_gravity = CenterGravity,
                            .win_gravity = StaticGravity,
                            .backing_store = Always, // Testing Always variant
                            .backing_planes = 0xFFFFFF,
                            .backing_pixel = 0xABCDEF,
                            .save_under = False,
                            .colormap = 333,
                            .map_installed = False,
                            .map_state = IsViewable,
                            .all_event_masks = 0,
                            .your_event_mask = 0,
                            .do_not_propagate_mask = 0,
                            .override_redirect = False,
                            .screen = (Screen *)0x3333};
  CAMLreturn(Val_xwindowattributes(&attr));
}

// 9. Unviewable State: A mapped window obscured by others
CAMLprim value get_xwa_unviewable(value unit) {
  CAMLparam1(unit);
  XWindowAttributes attr = {.x = 0,
                            .y = 0,
                            .width = 1920,
                            .height = 1080,
                            .border_width = 0,
                            .depth = 24,
                            .visual = (Visual *)0x4444,
                            .root = 444,
                            .class = InputOutput,
                            .bit_gravity = NorthWestGravity,
                            .win_gravity = NorthWestGravity,
                            .backing_store = WhenMapped,
                            .backing_planes = ~0UL,
                            .backing_pixel = 0,
                            .save_under = False,
                            .colormap = 444,
                            .map_installed = True,
                            .map_state = IsUnviewable,
                            .all_event_masks = 0,
                            .your_event_mask = 0,
                            .do_not_propagate_mask = 0,
                            .override_redirect = False,
                            .screen = (Screen *)0x4444};
  CAMLreturn(Val_xwindowattributes(&attr));
}

// 10. Max Pointer Values: Testing memory address upper limits
CAMLprim value get_xwa_max_pointers(value unit) {
  CAMLparam1(unit);
  XWindowAttributes attr = {
      .x = 0,
      .y = 0,
      .width = 1,
      .height = 1,
      .border_width = 0,
      .depth = 1,
      .visual = (Visual *)~0UL, // Max pointer bounds
      .root = ~0UL,
      .class = InputOutput,
      .bit_gravity = NorthWestGravity,
      .win_gravity = NorthWestGravity,
      .backing_store = NotUseful,
      .backing_planes = 0,
      .backing_pixel = 0,
      .save_under = False,
      .colormap = ~0UL, // Max XID
      .map_installed = False,
      .map_state = IsUnmapped,
      .all_event_masks = 0,
      .your_event_mask = 0,
      .do_not_propagate_mask = 0,
      .override_redirect = False,
      .screen = (Screen *)~0UL // Max pointer bounds
  };
  CAMLreturn(Val_xwindowattributes(&attr));
}

File: test/test_xwindowattributes.ml
open X11

(* External C functions *)
external get_xwa_basic : unit -> windowAttributes = "get_xwa_basic"
external get_xwa_zeroed : unit -> windowAttributes = "get_xwa_zeroed"

external get_xwa_negative_coords : unit -> windowAttributes
  = "get_xwa_negative_coords"

external get_xwa_huge_dimensions : unit -> windowAttributes
  = "get_xwa_huge_dimensions"

external get_xwa_input_only : unit -> windowAttributes = "get_xwa_input_only"

external get_xwa_max_masks_planes : unit -> windowAttributes
  = "get_xwa_max_masks_planes"

external get_xwa_bools_true : unit -> windowAttributes = "get_xwa_bools_true"

external get_xwa_gravities_store : unit -> windowAttributes
  = "get_xwa_gravities_store"

external get_xwa_unviewable : unit -> windowAttributes = "get_xwa_unviewable"

external get_xwa_max_pointers : unit -> windowAttributes
  = "get_xwa_max_pointers"

let test_xwa_basic () =
  let xwa = get_xwa_basic () in
  Alcotest.(check (pair int int)) "pos" (100, 200) xwa.pos;
  Alcotest.(check (pair int int)) "size" (800, 600) xwa.size;
  Alcotest.(check int) "border_width" 2 xwa.border_width;
  Alcotest.(check int) "depth" 24 xwa.depth;
  Alcotest.(check int) "c_class" 1 xwa.c_class;
  Alcotest.(check int) "bit_gravity" 1 xwa.bit_gravity;
  Alcotest.(check int) "win_gravity" 1 xwa.win_gravity;
  Alcotest.(check int) "backing_store" 0 xwa.backing_store;
  Alcotest.(check int) "backing_planes" 0xFFFFFFFF xwa.backing_planes;
  Alcotest.(check int) "backing_pixel" 0 xwa.backing_pixel;
  Alcotest.(check bool) "save_under" false xwa.save_under;
  Alcotest.(check int) "map_state" 2 xwa.map_state;
  Alcotest.(check bool)
    "all_event_masks" true
    (structureNotifyMask ||| exposureMask = xwa.all_event_masks);
  Alcotest.(check bool)
    "your_event_mask" true
    (structureNotifyMask = xwa.your_event_mask);
  Alcotest.(check int) "do_not_propagate_mask" 0 xwa.do_not_propagate_mask;
  Alcotest.(check bool) "override_redirect" false xwa.override_redirect;
  ()

let test_xwa_zeroed () =
  let xwa = get_xwa_zeroed () in
  Alcotest.(check (pair int int)) "pos" (0, 0) xwa.pos;
  Alcotest.(check (pair int int)) "size" (0, 0) xwa.size;
  Alcotest.(check int) "border_width" 0 xwa.border_width;
  Alcotest.(check int) "depth" 0 xwa.depth;
  Alcotest.(check int) "c_class" 0 xwa.c_class;
  Alcotest.(check int) "bit_gravity" 0 xwa.bit_gravity;
  Alcotest.(check int) "win_gravity" 0 xwa.win_gravity;
  Alcotest.(check int) "backing_store" 0 xwa.backing_store;
  Alcotest.(check int) "backing_planes" 0 xwa.backing_planes;
  Alcotest.(check int) "backing_pixel" 0 xwa.backing_pixel;
  Alcotest.(check bool) "save_under" false xwa.save_under;
  Alcotest.(check int) "map_state" 0 xwa.map_state;
  Alcotest.(check bool)
    "all_event_masks" true
    (noEventMask = xwa.all_event_masks);
  Alcotest.(check bool)
    "your_event_mask" true
    (noEventMask = xwa.your_event_mask);
  Alcotest.(check int) "do_not_propagate_mask" 0 xwa.do_not_propagate_mask;
  Alcotest.(check bool) "override_redirect" false xwa.override_redirect;
  ()

let test_xwa_negative_coords () =
  let xwa = get_xwa_negative_coords () in
  Alcotest.(check (pair int int)) "pos" (-1500, -2000) xwa.pos;
  Alcotest.(check (pair int int)) "size" (1920, 1080) xwa.size;
  Alcotest.(check int) "border_width" 0 xwa.border_width;
  Alcotest.(check int) "depth" 24 xwa.depth;
  Alcotest.(check int) "c_class" 1 xwa.c_class;
  Alcotest.(check int) "bit_gravity" 1 xwa.bit_gravity;
  Alcotest.(check int) "win_gravity" 1 xwa.win_gravity;
  Alcotest.(check int) "backing_store" 0 xwa.backing_store;
  Alcotest.(check bool) "save_under" false xwa.save_under;
  Alcotest.(check int) "map_state" 2 xwa.map_state;
  Alcotest.(check bool) "map_installed" true xwa.map_installed;
  Alcotest.(check bool) "override_redirect" false xwa.override_redirect;
  ()

let test_xwa_huge_dimensions () =
  let xwa = get_xwa_huge_dimensions () in
  Alcotest.(check (pair int int)) "pos" (0, 0) xwa.pos;
  Alcotest.(check (pair int int)) "size" (2147483647, 2147483647) xwa.size;
  Alcotest.(check int) "border_width" 2147483647 xwa.border_width;
  Alcotest.(check int) "depth" 32 xwa.depth;
  Alcotest.(check int) "c_class" 1 xwa.c_class;
  Alcotest.(check int) "bit_gravity" 1 xwa.bit_gravity;
  Alcotest.(check int) "win_gravity" 1 xwa.win_gravity;
  Alcotest.(check int) "backing_store" 0 xwa.backing_store;
  Alcotest.(check bool) "save_under" false xwa.save_under;
  Alcotest.(check int) "map_state" 2 xwa.map_state;
  Alcotest.(check bool) "map_installed" true xwa.map_installed;
  Alcotest.(check bool) "override_redirect" false xwa.override_redirect;
  ()

let test_xwa_input_only () =
  let xwa = get_xwa_input_only () in
  Alcotest.(check (pair int int)) "pos" (10, 10) xwa.pos;
  Alcotest.(check (pair int int)) "size" (100, 100) xwa.size;
  Alcotest.(check int) "border_width" 0 xwa.border_width;
  Alcotest.(check int) "depth" 0 xwa.depth;
  Alcotest.(check int) "c_class" 2 xwa.c_class;
  Alcotest.(check int) "bit_gravity" 0 xwa.bit_gravity;
  Alcotest.(check int) "win_gravity" 1 xwa.win_gravity;
  Alcotest.(check int) "backing_store" 0 xwa.backing_store;
  Alcotest.(check int) "backing_planes" 0 xwa.backing_planes;
  Alcotest.(check int) "backing_pixel" 0 xwa.backing_pixel;
  Alcotest.(check bool) "save_under" false xwa.save_under;
  (* Alcotest.(check int) "colormap" 0 xwa.colormap;*)
  Alcotest.(check bool) "map_installed" false xwa.map_installed;
  Alcotest.(check int) "map_state" 0 xwa.map_state;
  Alcotest.(check bool)
    "all_event_masks" true
    (enterWindowMask ||| leaveWindowMask = xwa.all_event_masks);
  Alcotest.(check bool)
    "your_event_mask" true
    (enterWindowMask = xwa.your_event_mask);
  Alcotest.(check bool) "override_redirect" true xwa.override_redirect;
  ()

let test_xwa_max_masks_planes () =
  let xwa = get_xwa_max_masks_planes () in
  Alcotest.(check (pair int int)) "pos" (0, 0) xwa.pos;
  Alcotest.(check (pair int int)) "size" (10, 10) xwa.size;
  Alcotest.(check int) "border_width" 0 xwa.border_width;
  Alcotest.(check int) "depth" 24 xwa.depth;
  Alcotest.(check int) "c_class" 1 xwa.c_class;
  Alcotest.(check int) "bit_gravity" 1 xwa.bit_gravity;
  Alcotest.(check int) "win_gravity" 1 xwa.win_gravity;
  Alcotest.(check int) "backing_store" 0 xwa.backing_store;
  Alcotest.(check int) "backing_planes max" (-1) xwa.backing_planes;
  Alcotest.(check int) "backing_pixel max" (-1) xwa.backing_pixel;
  Alcotest.(check bool) "save_under" false xwa.save_under;
  Alcotest.(check int) "map_state" 2 xwa.map_state;
  Alcotest.(check bool) "map_installed" true xwa.map_installed;
  (*Alcotest.(check bool)
    "all_event_masks max" true
    (xwa.all_event_masks = max_int);
  Alcotest.(check bool)
    "your_event_mask max" true
    (xwa.your_event_mask = max_int);*)
  Alcotest.(check int)
    "do_not_propagate_mask max" (-1) xwa.do_not_propagate_mask;
  Alcotest.(check bool) "override_redirect" false xwa.override_redirect;
  ()

let test_xwa_bools_true () =
  let xwa = get_xwa_bools_true () in
  Alcotest.(check (pair int int)) "pos" (5, 5) xwa.pos;
  Alcotest.(check (pair int int)) "size" (50, 50) xwa.size;
  Alcotest.(check int) "border_width" 1 xwa.border_width;
  Alcotest.(check int) "depth" 8 xwa.depth;
  Alcotest.(check int) "c_class" 1 xwa.c_class;
  Alcotest.(check int) "bit_gravity" 1 xwa.bit_gravity;
  Alcotest.(check int) "win_gravity" 1 xwa.win_gravity;
  Alcotest.(check int) "backing_store" 0 xwa.backing_store;
  Alcotest.(check int) "backing_planes" 0xFF xwa.backing_planes;
  Alcotest.(check int) "backing_pixel" 0xFF xwa.backing_pixel;
  Alcotest.(check bool) "save_under" true xwa.save_under;
  Alcotest.(check bool) "map_installed" true xwa.map_installed;
  Alcotest.(check int) "map_state" 2 xwa.map_state;
  Alcotest.(check int) "do_not_propagate_mask" 0 xwa.do_not_propagate_mask;
  Alcotest.(check bool) "override_redirect" true xwa.override_redirect;
  ()

let test_xwa_gravities_store () =
  let xwa = get_xwa_gravities_store () in
  Alcotest.(check (pair int int)) "pos" (20, 20) xwa.pos;
  Alcotest.(check (pair int int)) "size" (200, 200) xwa.size;
  Alcotest.(check int) "border_width" 5 xwa.border_width;
  Alcotest.(check int) "depth" 32 xwa.depth;
  Alcotest.(check int) "c_class" 1 xwa.c_class;
  Alcotest.(check int) "bit_gravity" 5 xwa.bit_gravity;
  Alcotest.(check int) "win_gravity" 10 xwa.win_gravity;
  Alcotest.(check int) "backing_store" 2 xwa.backing_store;
  Alcotest.(check int) "backing_planes" 0xFFFFFF xwa.backing_planes;
  Alcotest.(check int) "backing_pixel" 0xABCDEF xwa.backing_pixel;
  Alcotest.(check bool) "save_under" false xwa.save_under;
  Alcotest.(check int) "map_state" 2 xwa.map_state;
  Alcotest.(check bool) "map_installed" false xwa.map_installed;
  Alcotest.(check int) "do_not_propagate_mask" 0 xwa.do_not_propagate_mask;
  Alcotest.(check bool) "override_redirect" false xwa.override_redirect;
  ()

let test_xwa_unviewable () =
  let xwa = get_xwa_unviewable () in
  Alcotest.(check (pair int int)) "pos" (0, 0) xwa.pos;
  Alcotest.(check (pair int int)) "size" (1920, 1080) xwa.size;
  Alcotest.(check int) "border_width" 0 xwa.border_width;
  Alcotest.(check int) "depth" 24 xwa.depth;
  Alcotest.(check int) "c_class" 1 xwa.c_class;
  Alcotest.(check int) "bit_gravity" 1 xwa.bit_gravity;
  Alcotest.(check int) "win_gravity" 1 xwa.win_gravity;
  Alcotest.(check int) "backing_store" 1 xwa.backing_store;
  Alcotest.(check bool) "save_under" false xwa.save_under;
  Alcotest.(check bool) "map_installed" true xwa.map_installed;
  Alcotest.(check int) "map_state" 1 xwa.map_state;
  Alcotest.(check int) "do_not_propagate_mask" 0 xwa.do_not_propagate_mask;
  Alcotest.(check bool) "override_redirect" false xwa.override_redirect;
  ()

let test_xwa_max_pointers () =
  let xwa = get_xwa_max_pointers () in
  Alcotest.(check (pair int int)) "pos" (0, 0) xwa.pos;
  Alcotest.(check (pair int int)) "size" (1, 1) xwa.size;
  Alcotest.(check int) "border_width" 0 xwa.border_width;
  Alcotest.(check int) "depth" 1 xwa.depth;
  Alcotest.(check int) "c_class" 1 xwa.c_class;
  Alcotest.(check int) "bit_gravity" 1 xwa.bit_gravity;
  Alcotest.(check int) "win_gravity" 1 xwa.win_gravity;
  Alcotest.(check int) "backing_store" 0 xwa.backing_store;
  Alcotest.(check int) "backing_planes" 0 xwa.backing_planes;
  Alcotest.(check int) "backing_pixel" 0 xwa.backing_pixel;
  Alcotest.(check bool) "save_under" false xwa.save_under;
  (*Alcotest.(check int) "colormap max" max_int xwa.colormap;*)
  Alcotest.(check bool) "map_installed" false xwa.map_installed;
  Alcotest.(check int) "map_state" 0 xwa.map_state;
  Alcotest.(check bool)
    "all_event_masks" true
    (noEventMask = xwa.all_event_masks);
  Alcotest.(check bool)
    "your_event_mask" true
    (noEventMask = xwa.your_event_mask);
  Alcotest.(check int) "do_not_propagate_mask" 0 xwa.do_not_propagate_mask;
  Alcotest.(check bool) "override_redirect" false xwa.override_redirect;
  ()

let suite =
  let open Alcotest in
  [
    test_case "basic" `Quick test_xwa_basic;
    test_case "zeroed" `Quick test_xwa_zeroed;
    test_case "negative_coords" `Quick test_xwa_negative_coords;
    test_case "huge_dimensions" `Quick test_xwa_huge_dimensions;
    test_case "input_only" `Quick test_xwa_input_only;
    test_case "max_masks_planes" `Quick test_xwa_max_masks_planes;
    test_case "bools_true" `Quick test_xwa_bools_true;
    test_case "gravities_store" `Quick test_xwa_gravities_store;
    test_case "unviewable" `Quick test_xwa_unviewable;
    test_case "max_pointers" `Quick test_xwa_max_pointers;
  ]
